android/examples/todo-sync/src/main/java/com/mongodb/stitch/android/examples/todosync/TodoListActivity.java

```diff
@@ -136,26 +128,70 @@ public void updateTask(final ObjectId itemId, final String currentTask) {
           }
         });
     todoRecyclerView.setAdapter(todoAdapter);
+
+    doLogin();
+  }
+
+  private void doLogin() {
+    Stitch.getDefaultAppClient().getAuth().loginWithCredential(
+        new ServerApiKeyCredential(
+            "xEfxAP4jFWaWEs5WWpff7XyQMh1T56CCMmDEV9oxXtItPHBveA6bc6IEjOhQLes6"))
+        .addOnSuccessListener(user -> {
+          userId = user.getId();
+          invalidateOptionsMenu();
+          Toast.makeText(TodoListActivity.this, "Logged in", Toast.LENGTH_SHORT).show();
+          todoAdapter.replaceItems(getItemsFromServer(), false);
+
+          if (lists.sync().getSyncedIds().isEmpty()) {
+            lists.sync().insertOneAndSync(
+                new BsonDocument("_id", new BsonString(userId)));
+          }
+        })
+        .addOnFailureListener(e -> {
+          invalidateOptionsMenu();
+          Log.d(TAG, "error logging in", e);
+          Toast.makeText(TodoListActivity.this, "Failed logging in", Toast.LENGTH_SHORT).show();
+        });
+    System.out.println("RETURNING FROM LOGIN");
   }
 
   private class ListUpdateListener implements ChangeEventListener<BsonDocument> {
     @Override
     public void onEvent(final BsonValue documentId, final ChangeEvent<BsonDocument> event) {
       if (!event.hasUncommittedWrites()) {
-        todoAdapter.updateItems(getItemsFromServer());
+        todoAdapter.replaceItems(getItemsFromServer(), false);
       }
     }
   }
 
   private class ItemUpdateListener implements ChangeEventListener<Document> {
     @Override
     public void onEvent(final BsonValue documentId, final ChangeEvent<Document> event) {
-      if (!event.hasUncommittedWrites()) {
-        todoAdapter.updateItems(getItems());
+      final Handler mainHandler = new Handler(Looper.getMainLooper());
+      if (event.getOperationType() == ChangeEvent.OperationType.DELETE) {
+        mainHandler.post(() ->
```

Comment:
[minor] Not running on the main thread is going to throw many users off. We should other call all event listeners on the main thread or heavily document this.

Reply:
good point. So Tasks run their handlers on the main thread. It might be good for us to default to the main thread for this. Do you know if running event based actions tends to be on the main thread? I think the answer is yes given the event based system for Android

Reply:
Yeah I agree we should run on the main thread, especially if we're gonna do the same on iOS.